translator/trace/otlp/otlp_to_oc.go

```diff
@@ -0,0 +1,441 @@
+	if kindAttr := kindToAttribute(span.Kind); kindAttr != nil {
+		attributes.AttributeMap[tracetranslator.TagSpanKind] = kindAttr
+	}
+
+	return &octrace.Span{
+		TraceId:      span.TraceId,
+		SpanId:       span.SpanId,
+		Tracestate:   tracestateToOC(span.Tracestate),
+		ParentSpanId: span.ParentSpanId,
+		Name: &octrace.TruncatableString{
+			Value: span.Name,
+		},
+		Kind:           kindToOC(span.Kind),
+		StartTime:      unixnanoToTimestamp(span.StartTimeUnixnano),
+		EndTime:        unixnanoToTimestamp(span.EndTimeUnixnano),
+		Attributes:     attributes,
+		TimeEvents:     eventsToOC(span.Events, span.DroppedEventsCount),
+		Links:          linksToOC(span.Links, span.DroppedLinksCount),
+		Status:         statusToOC(span.Status),
+		ChildSpanCount: nil, // TODO(nilebox): Handle once OTLP supports it
+	}
+}
+
+func linksToOC(links []*otlptrace.Span_Link, droppedCount uint32) *octrace.Span_Links {
+	if len(links) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_Links{
+		Link:              linksToOCLinks(links),
+		DroppedLinksCount: int32(droppedCount),
+	}
+}
+
+func linksToOCLinks(links []*otlptrace.Span_Link) []*octrace.Span_Link {
+	if len(links) == 0 {
+		return nil
+	}
+
+	ocLinks := make([]*octrace.Span_Link, 0, len(links))
+	for _, link := range links {
+		ocLink := &octrace.Span_Link{
+			TraceId:    link.TraceId,
+			SpanId:     link.SpanId,
+			Tracestate: tracestateToOC(link.Tracestate),
+			Attributes: attributesToOCSpanAttributes(link.Attributes, link.DroppedAttributesCount),
+		}
+		ocLinks = append(ocLinks, ocLink)
+	}
+	return ocLinks
+}
+
+func eventsToOC(events []*otlptrace.Span_Event, droppedCount uint32) *octrace.Span_TimeEvents {
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 eventsToOCEvents(events),
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventsToOCEvents(events []*otlptrace.Span_Event) []*octrace.Span_TimeEvent {
+	if len(events) == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvent := &octrace.Span_TimeEvent{
+			Time: unixnanoToTimestamp(event.TimeUnixnano),
+		}
+
+		if isMessageEvent(event.Attributes) {
+			// Handle as a MessageEvent
+			ocEvent.Value = eventAttributesToMessageEvent(event.Attributes)
+		} else {
+			// Handle as an Annotation
+			spanAttributes := attributesToOCSpanAttributes(event.Attributes, event.DroppedAttributesCount)
+			ocEvent.Value = &octrace.Span_TimeEvent_Annotation_{
+				Annotation: &octrace.Span_TimeEvent_Annotation{
+					Description: &octrace.TruncatableString{
+						Value: event.Name,
+					},
+					Attributes: spanAttributes,
+				},
+			}
+		}
+
+		ocEvents = append(ocEvents, ocEvent)
+	}
+	return ocEvents
+}
+
+func isMessageEvent(attributes []*otlpcommon.AttributeKeyValue) bool {
+	for _, attr := range attributes {
+		if attr.Key == ocTimeEventMessageEventType {
```

Comment:
Not sure if there is a better way to distringuish message event from annotation? See `eventAttributesToMessageEvent` for specifics of reverse-mapping attributes to message event.

Reply:
There is no equivalent of MessageEvent in otel. Hence there is not need to derive something that is not there. IMO, events from otlp should be converted to Span_TimeEvent_Annotation.

Reply:
@tigrannajaryan can you confirm?
I did previously process all events as annotations, but then tried to apply reverse logic to `oc_to_otlp.go`.

Reply:
Will delete this code then.